apps/src/applab/manageAssets.jsx

```diff
@@ -0,0 +1,138 @@
+var React = require('react');
+var AssetsApi = require('./manageAssets/clientApi');
+var AssetRow = require('./manageAssets/assetRow.jsx');
+
+var errorMessages = {
+  415: 'This type of file is not supported.',
+  500: 'The server responded with an error.'
+};
+
+function getErrorMessage(status) {
+  return errorMessages[status] || 'An unknown error occurred.';
+}
+
+module.exports = React.createClass({
+  propTypes: {
+    assetChosen: React.PropTypes.func,
+    typeFilter: React.PropTypes.string
+  },
+
+  getInitialState: function () {
+    return {
+      assets: null,
+      uploadStatus: ''
+    };
+  },
+
+  componentWillMount: function () {
+    // TODO: Use Dave's client api when it's finished.
+    AssetsApi.ajax('GET', '', function (xhr) {
+      var assets = JSON.parse(xhr.responseText);
+      if (this.props.typeFilter) {
+        assets = assets.filter(function (asset) {
+          return asset.category === this.props.typeFilter;
+        }.bind(this));
+      }
+      this.setState({assets: assets});
+    }.bind(this), function (xhr) {
+      this.setState({uploadStatus: 'Error loading asset list: ' +
+          getErrorMessage(xhr.status)});
+    }.bind(this));
+  },
+
+  fileUploadClicked: function () {
+    var uploader = document.querySelector('#uploader');
+    uploader.click();
+  },
+
+  upload: function () {
+    var file = document.querySelector('#uploader').files[0];
+    var uploadStatus = document.querySelector('#uploadStatus');
+
+    if (file.type && this.props.typeFilter) {
+      var type = file.type.split('/')[0];
+      if (type !== this.props.typeFilter) {
+        this.setState({uploadStatus: 'Only ' + this.props.typeFilter +
+          ' assets can be used here.'});
+        return;
+      }
+    }
+
+    // TODO: Use Dave's client api when it's finished.
+    AssetsApi.ajax('PUT', file.name, function (xhr) {
+      this.state.assets.push(JSON.parse(xhr.responseText));
+      this.setState({uploadStatus: 'File "' + file.name +
+          '" successfully uploaded!'});
+    }.bind(this), function (xhr) {
+      this.setState({uploadStatus: 'Error uploading file: ' +
+          getErrorMessage(xhr.status)});
+    }.bind(this), file);
+
+    this.setState({uploadStatus: 'Uploading...'});
+  },
+
+  deleteAssetRow: function (name) {
+    this.setState({
+      assets: this.state.assets.filter(function (asset) {
+        return asset.filename !== name;
+      })
+    });
+    this.setState({uploadStatus: 'File "' + name + '" successfully deleted!'});
+  },
+
+  render: function () {
+    var assetList;
+    if (this.state.assets === null) {
```

Comment:
Might be worth a short comment explaining the difference between assets = null and assets.length = 0. (It looks like one means loading and one means no assets?)